### PR TITLE
Added Adam optimizer implementation

### PR DESCRIPTION
**Pull request by Spnetic-5**
#### Adam Implementation:

- Momentum Update:
`m = beta1 * m + (1 - beta1) * gradient`

- Adaptive Learning Rates Update:
`v = beta2 * v + (1 - beta2) * gradient**2`

- Bias-Corrected First and Second Moment Estimates:
`m_hat = m / (1 - beta1**t)`
`v_hat = v / (1 - beta2**t)`

- Parameter Update:
`param = param - learning_rate * m_hat / (sqrt(v_hat) + epsilon)`

### CODE REVIEW

File: src/nf/nf_optimizers.f90 [hunk 1]
```diff
@@ -123,4 +135,47 @@ pure subroutine minimize_rmsprop(self, param, gradient)
...
+    class(adam), intent(inout) :: self
+    integer, intent(in) :: num_params
+    if (.not. allocated(self % m)) then
+      allocate(self % m(num_params))
+      self % m = 0
+    end if
+    if (.not. allocated(self % v)) then
+      allocate(self % v(num_params))
+      self % v = 0
+    end if
```
Comment (milancurcic) [inline src/nf/nf_optimizers.f90:149]: `m` and `v` will always be either both allocated or deallocated so we can fold the second `if` block into the first one.

```suggestion
    if (.not. allocated(self % m)) then
      allocate(self % m(num_params), self % v(num_params))
      self % m = 0
      self % v = 0
    end if
```

File: src/nf/nf_optimizers.f90 [hunk 2]
```diff
@@ -59,6 +59,18 @@ end subroutine minimize
     procedure :: minimize => minimize_rmsprop
   end type rmsprop
 
+  type, extends(optimizer_base_type) :: adam
+    !! Adam optimizer
```
Comment (milancurcic) [inline src/nf/nf_optimizers.f90:63]: Let's add the reference to the paper here.

**Review by milancurcic — COMMENTED** [review]
Thank you! Good so far.

**Review by milancurcic — APPROVED** [review]
Thank you @Spnetic-5, this is almost good to go. The only thing missing is the convergence test for Adam in test/test_optimizers.f90.

File: src/nf/nf_optimizers.f90 [hunk 3]
```diff
@@ -123,4 +144,44 @@ pure subroutine minimize_rmsprop(self, param, gradient)
 
   end subroutine minimize_rmsprop
 
+
+  impure elemental subroutine init_adam(self, num_params)
+    class(adam), intent(inout) :: self
+    integer, intent(in) :: num_params
+    if (.not. allocated(self % m)) then
```
Comment (jvdp1) [inline src/nf/nf_optimizers.f90:151]: Could `v` be allocated and not `m`?
I suggest to check the status of both `m` and `v`.

Comment (milancurcic) [reply src/nf/nf_optimizers.f90:151]: That would be an error state, see https://github.com/modern-fortran/neural-fortran/pull/150#discussion_r1266806684. `v` and `m` will always have the same allocation status.

@Spnetic-5 let's make `m` and `v` `private` so that the user can't fiddle with them directly.

I'm also thinking whether we should make the `init` method `private` and invoke it from `minimize`. This decreases the number of places in the code at the network level that the optimizer instance is modified.

**Comment by milancurcic** [comment]
@Spnetic-5, just checking, your latest commit message mentions the convergence test but the commit doesn't actually include the test.

**Comment by Spnetic-5** [comment]
> @Spnetic-5, just checking, your latest commit message mentions the convergence test but the commit doesn't actually include the test.

Sorry, forgot to push test file😅️

File: src/nf/nf_optimizers.f90 [hunk 4]
```diff
@@ -123,4 +150,45 @@ pure subroutine minimize_rmsprop(self, param, gradient)
...
+    !! Concrete implementation of an Adam optimizer update rule.
+    class(adam), intent(inout) :: self
+    real, intent(inout) :: param(:)
+    real, intent(in) :: gradient(:)
+
+    self % t = self % t + 1
+
+    ! If weight_decay > 0, use L2 regularization;
+    ! otherwise, default to regular Adam.
+    associate(g => gradient + 2 * self % weight_decay * param)
```
Comment (milancurcic) [inline src/nf/nf_optimizers.f90:175]: Why do you have a factor of 2 here? In the [paper, Algorithm 2, line 6](https://arxiv.org/pdf/1711.05101.pdf) I see just $\lambda \theta_{t-1}$ (i.e. `self % weight_decay * param`).

Another question is, based on my understanding from the paper, applying `weight_decay * param` here is L2 regularization. However, for decoupled weight decay regularization, it seems like it should be added below on L188.

File: src/nf/nf_optimizers.f90 [hunk 5]
```diff
@@ -123,4 +150,45 @@ pure subroutine minimize_rmsprop(self, param, gradient)
...
+
+    ! Compute bias-corrected first and second moment estimates.
+    associate( &
+      m_hat => self % m / (1 - self % beta1**self % t), &
+      v_hat => self % v / (1 - self % beta2**self % t) &
+    )
+
+      ! Update parameters.
+      param = param - self % learning_rate &
+        * m_hat / (sqrt(v_hat) + self % epsilon)
```
Comment (milancurcic) [inline src/nf/nf_optimizers.f90:188]: Following up on my comment above, should the decoupled weight regularization not be:

```suggestion
      param = param - self % learning_rate &
        * m_hat / (sqrt(v_hat) + self % epsilon) &
        - self % weight_decay * param
```

File: src/nf/nf_optimizers.f90 (hunk 4 above)
Comment (Spnetic-5) [reply src/nf/nf_optimizers.f90:175]: You're right, the scaling factor of 2 is not present in the paper. The reason was when the weight decay coefficient is large, the term `weight_decay * param` can become very large.

Please review the recent changes.

File: src/nf/nf_optimizers.f90 (hunk 5 above)
Comment (Spnetic-5) [reply src/nf/nf_optimizers.f90:188]: Yeah, have made that change, just pushed

File: src/nf/nf_optimizers.f90 (hunk 4 above)
Comment (milancurcic) [reply src/nf/nf_optimizers.f90:175]: Thanks @Spnetic-5. We still need to decide how to handle the two different regularization approaches (L2 and decoupled). We currently have both enabled if `weight_decay > 0`. However, I think only one or the other should be applied at a time. PyTorch and Keras differentiate the two by providing two separate classes: Adam and AdamW. I think this is unnecessary in our case. We can have just Adam for now, and allow two different regularization approaches to be applied via input arguments. What do you think about using two separate weight decays to control this, i.e.:

```fortran
  real :: weight_decay_l2 = 0  ! L2 regularization (Adam)
  real :: weight_decay_decoupled = 0 ! decoupled weight decay regularization (AdamW)
```

Comment (Spnetic-5) [reply src/nf/nf_optimizers.f90:175]: Yup, I agree that having two separate input `weight_decay` arguments isa good approach. So by default, both weight decays will be set to 0, so no regularization will be applied. I'll go ahead and do the changes.

**Comment by milancurcic** [comment]
Thank you, @Spnetic-5, will merge. Please release v0.14.0 when ready.
